File: Code/Factory.Java/src/factory/secondversion/TheFactory.java
package factory.secondversion;

import factory.entities.ConcreteThing2;
import factory.interfaces.IThing;

/**
 * Factory that creates objects of the ConcreteThing2 class
 */
public class TheFactory extends BaseFactory {

	@Override
	public IThing createThing() {
		return new ConcreteThing2();
	}
}

File: Code/Factory.dotNet/Factory.dotNet/SecondVersion/TheFactory.cs
using Factory.Interfaces;
using Factory.Models;

namespace Factory.SecondVersion
{
    /// <summary>
    /// Factory that creates objects of the ConcreteThing2 class if some condition is met
    /// </summary>
    public class TheFactory : BaseFactory
    {
        public override IThing CreateThing()
        {
            bool someCondition = true;

            // If condition is met, then use the ConcreteThing2 class,
            // else use the ConcreteClass

            if (someCondition)
                return new ConcreteThing2();

            return base.CreateThing();
        }
    }
}
